### PR TITLE
ros timestamp fix (livox_lidar.cpp)

### PR DESCRIPTION
**Pull request by TobiMiller**
For me the timestamp was wrong and not useable with other publishers like TF.

### CODE REVIEW

**Comment by Livox-SDK** [comment]
Hi,TobiMiller
   The current timestamp is the absolute time since power-on,  the raw time of pointcloud.

**Comment by TobiMiller** [comment]
Does it make sense? When you do transforms ros will output errors because the time is not in range.
But doesn't matter. For me the changes fixed a problem.

**Comment by msadowski** [comment]
@TobiMiller makes a great point. Machine time always needs to be used as a timestamp otherwise most of the reasonable ROS packages will discard the data as it is too old

**Comment by Livox-SDK** [comment]
TobiMiller & msadowski ，
    In the next version, you can choose to use one of these two timestamps via the configuration file.
    The raw timestamp of pointcloud is usefully，when you need time synchronization with other lidars or sensors.
   Thanks a lot!

**Comment by RGring** [comment]
Is it by now possible to get pointclouds published with timestamps from the ROS time? (ros::Time::now())

**Comment by zhangmaoxiansheng** [comment]
> TobiMiller & msadowski ，
> In the next version, you can choose to use one of these two timestamps via the configuration file.
> The raw timestamp of pointcloud is usefully，when you need time synchronization with other lidars or sensors.
> Thanks a lot!

Using the latest version, how can I get the raw timestamp of pointcloud via the configuration file?

**Comment by YushengWHU** [comment]
> > TobiMiller & msadowski ，
> > In the next version, you can choose to use one of these two timestamps via the configuration file.
> > The raw timestamp of pointcloud is usefully，when you need time synchronization with other lidars or sensors.
> > Thanks a lot!
> 
> Using the latest version, how can I get the raw timestamp of pointcloud via the configuration file?

sure , still not solved

**Comment by borongyuan** [comment]
There is no need to modify the code. You did not provide a PTP master for Livox Lidar. You should try 'ptpd' or 'linuxptp', then the livox msg will have the correct timestamp and be perfectly sync with your host.

**Comment by VisionaryMind** [comment]
> There is no need to modify the code. You did not provide a PTP master for Livox Lidar. You should try 'ptpd' or 'linuxptp', then the livox msg will have the correct timestamp and be perfectly sync with your host.

Livox LiDAR is not capable of synchronizing with PTP masters. There is mention of this capability in the documentation, but at least as far as the Avia is concerned, regardless of whether or not a PTP master is present on the network with the Livox, it will never update its timestamp header to reflect the master clock's time. 

```
sudo ptp4l -i enp61s0 -l 6 -m

ptp4l[41927.906]: selected /dev/ptp0 as PTP clock
ptp4l[41927.907]: port 1: INITIALIZING to LISTENING on INIT_COMPLETE
ptp4l[41927.908]: port 0: INITIALIZING to LISTENING on INIT_COMPLETE
ptp4l[41935.189]: port 1: LISTENING to MASTER on ANNOUNCE_RECEIPT_TIMEOUT_EXPIRES
ptp4l[41935.189]: selected local clock 2cf05d.fffe.b18fb5 as best master
ptp4l[41935.189]: assuming the grand master role
```
Once that has been established, the Livox ROS driver is initialized. Here is a sample of its timestamp:

```
header: 
  seq: 1764
  stamp: 
    secs: 63
    nsecs: 549817080
  frame_id: "livox_frame"
height: 1
width: 7200
```
In fact, even the documentation on this repository states very clearly that the ROS driver does not support PTP synchronization: 

`    (1) The time stamp synchronization function of livox_ros_driver is based on the LidarSetUtcSyncTime interface of Livox-SDK, and only supports GPS synchronization, which is one of many synchronization methods of livox devices.`

In other words --- Livox ROS driver only supports GPS. It clearly does not work with PTP. My understanding, when I attempted this sync experiment, was that Livox devices synchronize automatically with PTP masters as soon as they are on the same network. This does not appear to be the case.

**Comment by borongyuan** [comment]
> Livox LiDAR is not capable of synchronizing with PTP masters, particularly when using the ROS driver. There is mention of this capability in the documentation, but at least as far as the Avia is concerned, regardless of whether or not a PTP master is present on the network with the Livox, it will never update its timestamp header to reflect the master clock's time.
PTP synchronization is not managed by Livox driver. or ROS driver, but it's supported. I've tested both ptpd and linuxptp. You will get the correct timestamp if you use one of them correctly.
> Here is an example of the master PTP initialization using linuxptp:
> 
> ```
> sudo ptp4l -i enp61s0 -l 6 -m
> 
> ptp4l[41927.906]: selected /dev/ptp0 as PTP clock
> ptp4l[41927.907]: port 1: INITIALIZING to LISTENING on INIT_COMPLETE
> ptp4l[41927.908]: port 0: INITIALIZING to LISTENING on INIT_COMPLETE
> ptp4l[41935.189]: port 1: LISTENING to MASTER on ANNOUNCE_RECEIPT_TIMEOUT_EXPIRES
> ptp4l[41935.189]: selected local clock 2cf05d.fffe.b18fb5 as best master
> ptp4l[41935.189]: assuming the grand master role
> ```
You only use ptp4l, which only synchronizes Livox and ptp0. But it doesn't synchronize ptp0 and system clock. You should also use phc2sys to synchronize them.
```
phc2sys -w -c enp61s0 -s CLOCK_REALTIME
```
It you use ptpd, it will automaticly synchronize HW clock and system clock.
